mlflow/java/client/src/main/java/org/mlflow/client/ApiClient.java

```diff
@@ -0,0 +1,140 @@
+package org.mlflow.client;
+
+import org.apache.http.client.utils.URIBuilder;
+
+import org.mlflow.api.proto.Service.*;
+import org.mlflow.client.creds.BasicMlflowHostCreds;
+import org.mlflow.client.creds.MlflowHostCredsProvider;
+import org.mlflow.client.objects.FromProtobufMapper;
+import org.mlflow.client.objects.ToProtobufMapper;
+
+import java.net.URI;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Client to an MLflow Tracking Sever.
+ */
+public class ApiClient {
+  private ToProtobufMapper toMapper = new ToProtobufMapper();
+  private FromProtobufMapper fromMapper = new FromProtobufMapper();
+  private HttpCaller httpCaller;
+
+  /**
+   * Creates a new ApiClient. Users should prefer constructing ApiClients via
+   * {@link #defaultClient()} or {@link #fromTrackingUri(String)} if possible.
+   */
+  public ApiClient(MlflowHostCredsProvider hostCredsProvider) {
+    httpCaller = new HttpCaller(hostCredsProvider);
+  }
+
+  /** Returns a default client based on the MLFLOW_TRACKING_URI environment variable. */
+  public static ApiClient defaultClient() {
```

Comment:
This is a pretty long thing to type. Should we just add a no-argument constructor, since we made the constructor public too? Then we could also make fromTrackingUri be a constructor with a string argument. Let's keep this super simple to initialize.